actors/common/src/main/java/org/sunbird/learner/actors/bulkupload/TextbookTocActor.java

```diff
@@ -70,20 +96,238 @@ public void onReceive(Request request) throws Throwable {
+    resultMap.remove(JsonKey.DIAL_CODES);
+
+    Set<String> topics = (Set<String>) resultMap.get(JsonKey.TOPICS);
+    resultMap.remove(JsonKey.TOPICS);
+
+    String tbId = (String) request.get(TEXTBOOK_ID);
+    Map<String, Object> textBookdata = getTextbook(tbId);
+
+    validateTopics(topics, (String) textBookdata.get(JsonKey.FRAMEWORK));
+    validateDialCodesWithReservedDialCodes(dialCodes, textBookdata);
+    request.getRequest().put(JsonKey.DATA, resultMap);
     String mode = ((Map<String, Object>) request.get(JsonKey.DATA)).get(JsonKey.MODE).toString();
-    validateRequest(request, mode);
+    validateRequest(request, mode, textBookdata);
     Response response = new Response();
-    if (StringUtils.equalsIgnoreCase(mode, "create")) {
-      response = createTextbook(request);
-    } else if (StringUtils.equalsIgnoreCase(mode, "update")) {
+    if (StringUtils.equalsIgnoreCase(mode, JsonKey.CREATE)) {
+      response = createTextbook(request, textBookdata);
+    } else if (StringUtils.equalsIgnoreCase(mode, JsonKey.UPDATE)) {
       response = updateTextbook(request);
     } else {
       unSupportedMessage();
     }
     sender().tell(response, sender());
   }
 
+  private void validateTopics(Set<String> topics, String frameworkId) {
+    Set<String> frameworkTopics = getRelatedFrameworkTopics(frameworkId);
+    topics.forEach(
+        name -> {
+          if (frameworkTopics.add(name)) {
+            throwClientErrorException(
+                ResponseCode.errorInvalidTopic,
+                MessageFormat.format(ResponseCode.errorInvalidTopic.getErrorMessage(), name));
+          }
+        });
+  }
+
+  @SuppressWarnings("unchecked")
+  private Set<String> getRelatedFrameworkTopics(String frameworkId) {
+    Response response = TextBookTocUtil.getRelatedFrameworkById(frameworkId);
+    Map<String, Object> result = response.getResult();
+    List<Map<String, Object>> terms = new ArrayList<>();
+    if (MapUtils.isNotEmpty(result)) {
+      Map<String, Object> framework = (Map<String, Object>) result.get(JsonKey.FRAMEWORK);
+      if (MapUtils.isNotEmpty(framework)) {
+        List<Map<String, Object>> categories =
+            (List<Map<String, Object>>) framework.get(JsonKey.CATEGORIES);
+        if (CollectionUtils.isNotEmpty(categories))
+          categories.forEach(
+              s -> {
+                if (JsonKey.TOPIC.equalsIgnoreCase((String) s.get(JsonKey.CODE))) {
+                  terms.addAll((List<Map<String, Object>>) s.get(JsonKey.TERMS));
+                }
+              });
+      }
+    }
+    Set<String> topics = new HashSet<>();
+    for (Map<String, Object> term : terms) {
+      List<Map<String, Object>> childrens = (List<Map<String, Object>>) term.get(JsonKey.CHILDREN);
+      if (CollectionUtils.isNotEmpty(childrens)) {
+        for (Map<String, Object> child : childrens) {
+          topics.add((String) child.get(JsonKey.NAME));
```

Comment:
topcis could be only till 2 levels - if it is recursive - this only takes care of 2 levels. Hence asking.